src/System.Reflection.Metadata/src/System/Reflection/PortableExecutable/PEReader.cs

```diff
@@ -716,5 +717,230 @@ internal static unsafe ImmutableArray<byte> DecodeEmbeddedPortablePdbDebugDirect
 
             return ImmutableByteArrayInterop.DangerousCreateFromUnderlyingArray(ref decompressed);
         }
+
+        /// <summary>
+        /// Opens a Portable PDB associated with this PE image.
+        /// </summary>
+        /// <param name="peImagePath">
+        /// The path to the PE image, or <c>null</c> if the image doesn't exist on disk.
+        /// The path is used to locate the PDB file next to the PE image.
+        /// </param>
+        /// <param name="pdbFileStreamProvider">
+        /// If specified, called to open a <see cref="Stream"/> for a given file path. 
+        /// The provider is expected to either return a readable and seekable <see cref="Stream"/>, 
+        /// or <c>null</c> if the target file doesn't exist or should be ignored for some reason.
+        /// 
+        /// If not specified, the CodeView entry is ignored (if present) and the Embedded Portable PDB is open (if present).
+        /// 
+        /// The provider shall throw <see cref="IOException"/> if it fails to open the file due to an unexpected IO error.
+        /// </param>
+        /// <returns>
+        /// A new instance of <see cref="MetadataReaderProvider"/> to be used to read the Portable PDB, 
+        /// or null if the PE image doesn't have a PDB associated with it of the PDB couldn't be read.
+        /// </returns>
+        /// <remarks>
+        /// Implements a simple PDB file lookup based on the content of the PE image Debug Directory.
+        /// A sophisticated tool might need to follow up with additional look up on search paths or symbol server.
+        /// 
+        /// The method looks the PDB up in the following steps in the listed order:
+        /// 1) Check for a matching PDB file of the name found in the CodeView entry in the directory containing the PE image (directory of <paramref name="peImagePath"/>).
+        /// 2) Check for a matching PDB file on the exact path found in the CodeView entry.
+        /// 3) Check for a PDB embedded in the PE image itself.
+        /// 
+        /// The first PDB that matches the information specified in the Debug Directory is returned.
+        /// </remarks>
+        /// <exception cref="InvalidOperationException">The stream returned from <paramref name="pdbFileStreamProvider"/> doesn't support read and seek operations.</exception>
+        /// <exception cref="BadImageFormatException">No matching PDB file is found due to an error: The PE image or the PDB is invalid.</exception>
+        /// <exception cref="IOException">No matching PDB file is found due to an error: An IO error occurred while reading the PE image or the PDB.</exception>
+        public MetadataReaderProvider OpenAssociatedPortablePdb(string peImagePath, Func<string, Stream> pdbFileStreamProvider)
```

Comment:
API looks good. Error handling seems right, but also quite tricky so we'll need a good set of tests.